ord_schema/process_dataset.py

```diff
@@ -0,0 +1,143 @@
+  $ python process_dataset.py --input_pattern=my_dataset.pb --write_errors
+* To process multiple Dataset protos:
+  $ python process_dataset.py --input_pattern="my_dataset-*.pb"
+"""
+
+import glob
+import os
+import uuid
+
+from absl import app
+from absl import flags
+from absl import logging
+
+from ord_schema import message_helpers
+from ord_schema import validations
+from ord_schema.proto import dataset_pb2
+
+FLAGS = flags.FLAGS
+flags.DEFINE_string('input_pattern', None,
+                    'Pattern (glob) matching input Dataset protos.')
+flags.DEFINE_enum('input_format', 'binary',
+                  [f.value for f in message_helpers.MessageFormats],
+                  'Input message format.')
+flags.DEFINE_boolean('write_errors', False,
+                     'If True, errors will be written to <filename>.error.')
+flags.DEFINE_string('output', None, 'Filename for output Dataset.')
+flags.DEFINE_boolean('validate', True, 'If True, validate Reaction protos.')
+flags.DEFINE_boolean('update', False, 'If True, update Reaction protos.')
+
+
+def validate(datasets):
+    """Runs validation for a set of datasets.
+
+    Args:
+        datasets: Dict mapping filenames to Dataset protos.
+    """
+    validation_errors = False
+    for filename, dataset in datasets.items():
+        errors = _validate_dataset(filename, dataset)
+        if errors:
+            validation_errors = True
+            if FLAGS.write_errors:
+                with open(f'{filename}.error', 'w') as f:
+                    for error in errors:
+                        f.write(f'{error}\n')
+    # NOTE(kearnes): Run validation for all datasets before
+    # exiting if there are errors.
+    if validation_errors:
+        raise ValueError('validation encountered errors')
+
+
+def _validate_dataset(filename, dataset):
+    """Validates Reaction messages in a Dataset.
+
+    Note that validation may change the message. For example, NAME
+    identifiers will be resolved to structures.
+
+    Args:
+        filename: Text filename; the dataset source.
+        dataset: dataset_pb2.Dataset message.
+
+    Returns:
+        List of validation error messages.
+    """
+    basename = os.path.basename(filename)
+    errors = []
+    for reaction in dataset.reactions:
+        for error in validations.validate_message(
+                reaction, raise_on_error=False):
+            errors.append(error)
+            logging.warning('Validation error for %s: %s', basename, error)
+    logging.info('Validation summary for %s: %d/%d successful (%d failures)',
+                 basename,
+                 len(dataset.reactions) - len(errors),
+                 len(dataset.reactions),
+                 len(errors))
+    return errors
+
+
+def _update_reaction(reaction):
+    """Updates a Reaction message.
+
+    Current updates:
+      * Sets record_id if not already set.
+
+    Args:
+        reaction: reaction_pb2.Reaction message.
+    """
+    if not reaction.provenance.record_id:
```

Comment:
Do we want to only set the `record_id` if not already set? What if a submitter puts in a dummy placeholder value or a malicious value? 

I suppose we would ideally check if the `record_id` was previously set _by this validation script_. Perhaps you can drop a TODO in here, since I think the logic we will eventually want will be more complex than `if not reaction.provenance.record_id`.

Reply:
Agreed there will be more complexity here; will add a TODO.